### PR TITLE
byron-proxy: Byron peers can sync

### PR DESCRIPTION
**Pull request by avieth**
- A Byron server can run concurrently with a Shelley client. The latter will be used as the sole block source (until we have the tools to deal with concurrent sources).
- The tip is announced to Byron peers every 10 seconds. Without this, they would never attempt to download from the proxy unless they were in the infamous "recovery mode". Support from a forthcoming block storage layer will be required to eliminate the polling style. Alternatively, we could bring in some slotting- and time-aware mechanism to announce on slot boundaries.
- Fixed a DB bug when appending a `cardano-ledger` block (as opposed to a `cardano-sl` block).
- Fixed bugs in the DB-backed `cardano-sl` `Logic` implementation where they did not match the intended contract reverse-engineered from the actual `cardano-sl` full logic layer implementation.
- Also includes better trace output formatting.

### CODE REVIEW

**Comment by avieth** [comment]
Bump. @dcoutts @coot need a review.

File: byron-proxy/src/lib/Ouroboros/Byron/Proxy/Main.hs [hunk 1]
```diff
@@ -345,48 +349,49 @@ bbsGetBlockHeaders db onErr mLimit checkpoints mTip = do
...
-  consumer n acc = do
-    next <- await
-    case next of
-      Nothing        -> pure Nothing
-      Just blkHeader ->
-        if maybe False ((==) (headerHash blkHeader)) mTip
-        then pure $ Just $ NewestFirst $ blkHeader NE.:| acc
-        else consumer (n-1) (blkHeader : acc)
+  consumer limit 0 acc = do
+    -- Drop the first one.
```
Comment (coot) [inline byron-proxy/src/lib/Ouroboros/Byron/Proxy/Main.hs]: why? is this the infamous ebb?

File: byron-proxy/src/exec/Logging.hs [hunk 2]
```diff
@@ -58,12 +60,11 @@ convertTrace trace = case trace of
...
 convertTrace'
-  :: Monitoring.Trace IO a
-  -> Tracer IO (Monitoring.LoggerName, Monitoring.Severity, a)
+  :: Monitoring.Trace IO Text
+  -> Tracer IO (Monitoring.LoggerName, Monitoring.Severity, Text.Builder)
 convertTrace' = contramap f . convertTrace
   where
-  f (a, b, c) = (a, b, Monitoring.LogMessage c)
-
+  f (a, b, builder) = (a, b, Monitoring.LogMessage (Text.toStrict (Text.toLazyText builder)))
```
[the remark this answers is not in the document]
Comment (avieth) [reply byron-proxy/src/exec/Logging.hs]: AFAICT the text builder can't go straight to strict text.

[the remark this answers is not in the document]
Comment (avieth) [reply byron-proxy/src/exec/Logging.hs]: I don't think it's an extra copy. I think it's just one copy for each buffer required by the builder.

File: byron-proxy/src/lib/Ouroboros/Byron/Proxy/Main.hs (hunk 1 above)
Comment (avieth) [reply byron-proxy/src/lib/Ouroboros/Byron/Proxy/Main.hs]: No, it's to keep with the de facto cardano-sl contract. The checkpoint itself is not included in the result. Comment on lines 334, 335 explains it.

**Pull request merged**